### PR TITLE
ACCUMULO-2232 Added options to Combiner for handling deletes

### PR DESCRIPTION
**Pull request by keith-turner**

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java [hunk 1]
```diff
@@ -149,6 +159,37 @@ public void next() throws IOException {
 
   private Key workKey = new Key();
 
+  private static Cache<String,Long> loggedMsgCache = CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.HOURS).build();
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: +final

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java [hunk 2]
```diff
@@ -227,10 +271,25 @@ public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> op
       throw new IllegalArgumentException("The " + COLUMNS_OPTION + " must not be empty");
 
     combiners = new ColumnSet(Lists.newArrayList(Splitter.on(",").split(encodedColumns)));
+
+    isPartialCompaction = ((env.getIteratorScope() == IteratorScope.majc) && !env.isFullMajorCompaction());
+    if (options.containsKey(DELETE_HANDLING_ACTION_OPTION)) {
+      deleteHandlingAction = DeleteHandlingAction.valueOf(options.get(DELETE_HANDLING_ACTION_OPTION));
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: A warning message with the actual property being checked (DELETE_HANDLING_ACTION_OPTION) and the value that we have no enum-value for would be useful.

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java [hunk 3]
```diff
@@ -240,15 +299,18 @@ public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> op
...
+    newInstance.deleteHandlingAction = deleteHandlingAction;
     return newInstance;
   }
 
   @Override
   public IteratorOptions describeOptions() {
     IteratorOptions io = new IteratorOptions("comb", "Combiners apply reduce functions to multiple versions of values with otherwise equal keys", null, null);
-    io.addNamedOption(ALL_OPTION, "set to true to apply Combiner to every column, otherwise leave blank. if true, " + COLUMNS_OPTION
-        + " option will be ignored.");
+    io.addNamedOption(ALL_OPTION,
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: Nit: unrelated formatting.

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java [hunk 4]
```diff
@@ -240,15 +299,18 @@ public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> op
...
+        "set to true to apply Combiner to every column, otherwise leave blank. if true, " + COLUMNS_OPTION + " option will be ignored.");
     io.addNamedOption(COLUMNS_OPTION, "<col fam>[:<col qual>]{,<col fam>[:<col qual>]} escape non-alphanum chars using %<hex>.");
+    // TODO
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: Did you forget todo something?

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java [hunk 5]
```diff
@@ -275,6 +337,10 @@ public boolean validateOptions(Map<String,String> options) {
         throw new IllegalArgumentException("invalid column encoding " + encodedColumns);
     }
 
+    if (options.containsKey(DELETE_HANDLING_ACTION_OPTION)) {
+      DeleteHandlingAction.valueOf(options.get(DELETE_HANDLING_ACTION_OPTION));
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: Same suggestion about a meaningful error message here.

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java [hunk 6]
```diff
@@ -313,4 +378,46 @@ public static void setColumns(IteratorSetting is, List<IteratorSetting.Column> c
...
+
+    /**
+     * Pass all data through during partial major compactions, no reducing is done. With this option reducing is only done during scan and full major
+     * compactions, when deletes can be correctly handled.
+     */
+    REDUCE_ON_FULL_COMPACTION_ONLY
+  }
+
+  /**
+   * Combiners may not work correctly with deletes. Sometimes when Accumulo compacts the files in a tablet, it only compacts a subset of the files. If a delete
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: Great writeup.

File: core/src/test/java/org/apache/accumulo/core/iterators/user/CombinerTest.java [hunk 7]
```diff
@@ -786,4 +816,107 @@ public void testAdds() {
...
+      }
+
+      ai.init(new SortedMapIterator(input), is.getOptions(), env);
+
+      if (deepCopy)
+        assertEquals(expected, readAll(ai.deepCopy(env)));
+      assertEquals(expected, readAll(ai));
+
+    } finally {
+      logger.removeAppender(appender);
```
Comment (joshelser) [inline core/src/test/java/org/apache/accumulo/core/iterators/user/CombinerTest.java]: Glad to see the try/finally logger reset.

File: core/src/test/java/org/apache/accumulo/core/iterators/user/CombinerTest.java [hunk 8]
```diff
@@ -786,4 +816,107 @@ public void testAdds() {
...
+    TreeMap<Key,Value> expected = new TreeMap<Key,Value>();
+    nkv(expected, 1, 1, 1, 1, false, 4l, encoder);
+    nkv(expected, 1, 1, 1, 2, true, 0l, encoder);
+    nkv(expected, 1, 1, 1, 4, false, 11l, encoder);
+
+    try {
+      runDeleteHandlingTest(input, expected, DeleteHandlingAction.THROW_EXCEPTION, paritalMajcIe);
+      Assert.fail();
+    } catch (IllegalStateException ise) {
+      Assert.assertTrue(ise.getMessage().contains("Saw a delete during a partial compaction"));
```
Comment (joshelser) [inline core/src/test/java/org/apache/accumulo/core/iterators/user/CombinerTest.java]: Try to include a meaningful error message if the assertion fails. We should be able to explain the exact reason the test failed just by looking at surefire output.

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java (hunk 1 above)
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: Would be safer to include a high-watermark on the size of the cache. Maybe 500 elements just to prevent catastrophic failure if we somehow create a bug that causes things to be added to this cache incorrectly.

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java [hunk 9]
```diff
@@ -149,6 +159,37 @@ public void next() throws IOException {
...
+    if (isPartialCompaction) {
+      switch (deleteHandlingAction) {
+        case LOG_ERROR:
+          try {
+            loggedMsgCache.get(this.getClass().getName(), new Callable<Long>() {
+              @Override
+              public Long call() throws Exception {
+                log.error("Combiner of type " + this.getClass().getSimpleName()
+                    + " saw a delete during a partial compaction.  This could cause undesired results.  See ACCUMULO-2232.  Will not log subsequent occurences for at least 1 hour.");
+                return System.currentTimeMillis();
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: Do we really need the current time? Isn't `0` sufficient since we don't inspect the value of the cache entries? A comment that the value returned by the Callable is meaningless would also be nice.

**Comment by joshelser** [comment]
As commented on JIRA before, approach looks good. Pointed out what I see as room for some improvements.

**Comment by keith-turner** [comment]
@joshelser thanks for the feedback, it was very useful.   I think I fixed all the issues you found.

**Comment by joshelser** [comment]
Great. I'll try to take another peek today, but if you think it's ready, don't feel obligated to wait on me.

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java [hunk 10]
```diff
@@ -149,6 +160,38 @@ public void next() throws IOException {
...
   private Key workKey = new Key();
 
+  private static final Cache<String,Long> loggedMsgCache = CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.HOURS).maximumSize(10000).build();
+
+  private void sawDelete() {
+    if (isPartialCompaction) {
+      switch (deleteHandlingAction) {
+        case LOG_ERROR:
+          try {
+            loggedMsgCache.get(this.getClass().getName(), new Callable<Long>() {
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: At first, I thought "Callable<Void>" might be better, but apparently CacheBuilder doesn't like that, so might be better to use "Callable<Boolean>", because it's more readable to see "Boolean.TRUE" inserted into the cache to denote that a log message has recently occurred. (Unless you were going to use the integer to denote how many times it has been logged before suppressing repeats after some threshold, like 5.)

**Comment by keith-turner** [comment]
In 31fcf00 I changed the behaviour slightly.

I dropped the option to throw an exception.  I was thinking this option could prevent problems, but it can not.  It can still only detect problems.  For detecting problems I think logging an error is sufficient.

I also changed logging to log on partial and full major compactions.   If a full major compaction sees a delete, then its evidence that a problem could have occurred in the past with a partial compaction.

**Comment by joshelser** [comment]
:+1:

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java [hunk 11]
```diff
@@ -313,4 +392,48 @@ public static void setColumns(IteratorSetting is, List<IteratorSetting.Column> c
   public static void setCombineAllColumns(IteratorSetting is, boolean combineAllColumns) {
     is.addOption(ALL_OPTION, Boolean.toString(combineAllColumns));
   }
+
+  /**
+   * @since 1.6.4 1.7.1 1.8.0
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: Instead of offering the options `IGNORE` and `LOG_ERROR`, we should probably just always log a warning (if `REDUCE_ON_FULL_COMPACTION_ONLY` is not set) and let users disable it in the log4j config, rather than in code.

The `REDUCE_ON_FULL_COMPACTION_ONLY` should be treated as a distinct boolean value which can be set to `true` or `false` as a normal iterator option (similar to `Filter`'s `negate` option), with a default value of `false` to preserve existing behavior.

That would also minimize API changes, especially for 1.6.x and 1.7.x. While this isn't technically public API, it'd still be good to minimize API impact on users using those versions.

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java [hunk 12]
```diff
@@ -313,4 +392,48 @@ public static void setColumns(IteratorSetting is, List<IteratorSetting.Column> c
...
+   * <a href="https://issues.apache.org/jira/browse/ACCUMULO-2232">ACCUMULO-2232</a> for more information.
+   *
+   * <p>
+   * This method allows users to configure how they want to handle the combination of delete markers, combiners, and major compactions. The default behavior is
+   * {@link DeleteHandlingAction#LOG_ERROR}. See the javadoc on each {@link DeleteHandlingAction} enum for a description of each option.
+   *
+   * <p>
+   * For correctness deletes should not be used with columns that are combined OR the {@link DeleteHandlingAction#REDUCE_ON_FULL_COMPACTION_ONLY} option should
+   * be used. Only reducing on full major compactions may have negative performance implications.
+   *
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: Another option to give users the ability to "delete" without using problematic deletes is to give them the ability to return a `null` value, with the semantics of `null` being "drop them all". This, combined with the ability to detect which iterator scope and and major compaction mode is being run, could be used to drop data in a sensible way, suitable to the user's schema, without inserting problematic delete markers.

If we did that, it'd be good to add that here, with the recommendation to avoid using deletes entirely.

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java (hunk 11 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: I like this approach because it simplified things.  In bbea55d I got rid of the enum and swithced to a boolean.  There is still an addition of a public method to Combiner.

File: core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java (hunk 12 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java]: I don't think this should be coupled with this fix.  Also not sure about doing this in 1.6.  I opened a new issue [ACCUMULO-4011](https://issues.apache.org/jira/browse/ACCUMULO-4011)
